### ISSUE TITLE
Excessive 'possible stale hash' messages

### ISSUE THREAD

**Issue by v-dobrev** [libocca/occa#199]
When running on multiple compute nodes, OCCA produces a large number of messages like this:
```
Located possible stale hash: [</path/to/occa/cache/locks/...>]
```

I assume that the reason behind these messages is that all MPI ranks are trying to lock the same file/directory causing slowdown and the hence the above messages. When running large MPI jobs, it seems appropriate to use MPI communications for synchronization.

Any thoughts on addressing this issue?

**Comment by v-dobrev** [comment]
In rare cases, the above messages can also be seen when running multiple MPI tasks on a single node.

**Comment by dmed256** [comment]
Ah, there's a stale warning after 10s of waiting and will replace the file after 20s

To change those defaults:
```
occa::properties &settings = occa::settings();
settings["locks/stale-warning"] = 40.0; // Warn after I've waited 40 seconds
settings["locks/stale-age"]     = 60.0; // Replace the lock after it's over 60 seconds old
```

If it's showing up a lot, we can change the defaults too

**Comment by v-dobrev** [comment]
I guess my main concern is not just the messages themselves but the slowdown in loading existing kernels from the cache when running multiple MPI tasks. The messages are mostly a consequence of that slowdown.

For example, when running 16 MPI tasks (each controlling 1 GPU) across 4 nodes, loading certain kernels can take more than 20 seconds. At the same time, loading the same kernels when running 4 MPI tasks on one node takes less than a second.

**Comment by dmed256** [comment]
### TL;DR
 It's getting added soon by prefetching cached kernels using `device::loadKernels()`

### Description
We saw this in Laghos and added 2 features:

- Kernel caching at the `device` level to avoid loading from cache after the first build
- `device::loadKernels` to cache already-built kernels

Turns out caching the kernels meant having to do additional checks the existing reference counting method didn't catch. Both were turned off when it was caught due to libCEED fails:

https://github.com/libocca/occa/blob/master/src/device.cpp#L509

In addition to keeping track of `kernel` objects using the same `modeKernel_t` object, the number of times the kernel was _created_ needs to be stored somewhere to avoid one `kernel::free` messing up all other instances. There's a second type of reference counting added a few days ago (https://github.com/libocca/occa/commit/58e47f5ccf0d87f5b91e6851b2d74a9456c46b2c) that I hope to merge with `kernel` soon :)
